### DISCUSSION TITLE
How to monitor Java applications with Nightingale

### DISCUSSION THREAD

**Discussion by IT-Forerunner** [ccfos/nightingale#1221] [Q&A]
我参看夜莺监控程序，是通过PID文件，和程序name来判断的。但是java程序的name都交java。也没有PID文件。尝试过categraf。他也是通过name来获取进程的。
请问如何能做到监控java程序，并且java程序更新或者重启都不影响监控。尽量不对java程序修改的方法。

**Comment by UlricQin** [comment]
使用 categraf 的 procstat 插件，使用 search_cmdline_substring 来做过滤，procstat.toml 里已经给了详尽注释，假设你的程序运行时，有个参数是 abc.properties，配置文件可以这么写：

```toml
[[instances]]
search_cmdline_substring = "abc.properties"
gather_total = true
gather_per_pid = false
gather_more_metrics = [
    "threads",
    "fd",
    "io",
    "uptime",
    "cpu",
    "mem",
    "limit"
]
```
